runtime/src/bank/tests.rs

```diff
@@ -8013,44 +8013,208 @@ fn test_compute_active_feature_set() {
     assert!(bank.feature_set.is_active(&test_feature));
 }
 
+fn set_up_account_with_bank(
+    bank: &mut Bank,
+    pubkey: &Pubkey,
+    lamports: u64,
+    data: Vec<u8>,
+) -> AccountSharedData {
+    let new_account = AccountSharedData::from(Account {
+        lamports,
+        data,
+        ..Account::default()
+    });
+    bank.store_account_and_update_capitalization(pubkey, &new_account);
+    assert_eq!(bank.get_balance(pubkey), lamports);
+    new_account
+}
+
 #[test]
-fn test_program_replacement() {
+fn test_non_upgradable_program_replace() {
+    // Non-Upgradable program
+    // - Old:     [Old program data]
+    // - New:     [*New program data]
+    //
+    // Should replace the old program account with the new program account:
+    // - Old:     [*New program data]
     let mut bank = create_simple_test_bank(0);
 
-    // Setup original program account
-    let old_address = Pubkey::new_unique();
-    let new_address = Pubkey::new_unique();
-    bank.store_account_and_update_capitalization(
-        &old_address,
-        &AccountSharedData::from(Account {
-            lamports: 100,
-            ..Account::default()
-        }),
-    );
-    assert_eq!(bank.get_balance(&old_address), 100);
-
-    // Setup new program account
-    let new_program_account = AccountSharedData::from(Account {
-        lamports: 123,
-        ..Account::default()
-    });
-    bank.store_account_and_update_capitalization(&new_address, &new_program_account);
-    assert_eq!(bank.get_balance(&new_address), 123);
+    let old = Pubkey::new_unique();
+    set_up_account_with_bank(&mut bank, &old, 100, vec![0, 0, 0, 0]);
+
+    let new = Pubkey::new_unique();
+    let new_program_bytes = vec![6, 5, 4, 3, 2, 1, 0];
+    set_up_account_with_bank(&mut bank, &new, 100, new_program_bytes.clone());
 
     let original_capitalization = bank.capitalization();
 
-    bank.replace_program_account(&old_address, &new_address, "bank-apply_program_replacement");
+    bank.replace_program_account(&old, &new, "bank-apply_program_replacement");
+
+    // Old program account balance is unchanged
+    assert_eq!(bank.get_balance(&old), 100);
```

Comment:
Shouldn't it have the balance of the new account? If we replace an old program with a new bigger program, the old address should now have the balance from the new address, no?